mlir/lib/Analysis/Presburger/Matrix.cpp

```diff
@@ -432,4 +432,111 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
 
 MPInt IntMatrix::normalizeRow(unsigned row) {
   return normalizeRow(row, getNumColumns());
+}
+
+MPInt IntMatrix::determinant(IntMatrix* inverse) const {
+  unsigned nRows = getNumRows();
+  unsigned nColumns = getNumColumns();
+
+  assert(nRows == nColumns && "determinant can only be calculated for square matrices!");
+
+  FracMatrix m(*this);
+  
+  FracMatrix fracInverseM(nRows, nColumns);
+  FracMatrix* fracInverse = &fracInverseM;
+  Fraction detF = m.determinant(fracInverse);
+
+  MPInt detM = detF.getAsInteger();
+
+  if (detM != 0) {
+    *inverse = IntMatrix(nRows, nColumns);
+    for (unsigned i = 0; i < nRows; i++)
+      for (unsigned j = 0; j < nColumns; j++)
+        inverse->at(i, j) = (fracInverse->at(i, j) * detF).getAsInteger();
+  }
+
+  return detM;
+}
+
+FracMatrix FracMatrix::identity(unsigned dimension) {
+  return Matrix::identity(dimension);
+}
+
+FracMatrix::FracMatrix(IntMatrix m)
+  : FracMatrix(m.getNumRows(), m.getNumColumns()) {
+  for (unsigned i = 0; i < m.getNumRows(); i++)
+    for (unsigned j = 0; j < m.getNumColumns(); j++)
+      this->at(i, j) = Fraction(m.at(i, j), 1);
+}
+
+Fraction FracMatrix::determinant(FracMatrix* inverse) const {
+  unsigned nRows = getNumRows();
+  unsigned nColumns = getNumColumns();
+
+  assert(nRows == nColumns && "determinant can only be calculated for square matrices!");
+
+  FracMatrix m(*this);
+  FracMatrix tempInv(nRows, nColumns);
+  if (inverse) tempInv = FracMatrix::identity(nRows);
+
+  Fraction a, b;
+  // For each row in the matrix,
+  for (unsigned i = 0; i < nRows; i++) {
+    if (m(i, i) == 0)
+    // First ensure that the diagonal
+    // element is nonzero, by adding
+    // a nonzero row to it.
+      for (unsigned j = i + 1; j < nRows; j++)
+        if (m(j, i) != 0) {
+          m.addToRow(j, i, 1);
+          if (inverse) tempInv.addToRow(j, i, 1);
+          break;
+        }
+
+    b = m.at(i, i);
+    if (b == 0)
+        return 0;
+    
+    if (inverse) {
+      for (unsigned j = 0; j < i; j++) {
+        if (m.at(j, i) == 0)
+          continue;
+        a = m.at(j, i);
+        // Set element (j, i) to zero
+        // by subtracting the ith row,
+        // appropriately scaled.
+        m.addToRow(i, j, -a / b);
+        tempInv.addToRow(i, j, -a / b);
+      }
+    }
+
+    // Set all elements below the
+    // diagonal to zero.
+    for (unsigned j = i+1; j < nRows; j++) {
+      if (m.at(j, i) == 0)
+        continue;
+      a = m.at(j, i);
+      // Set element (j, i) to zero
+      // by subtracting the ith row,
+      // appropriately scaled.
+      m.addToRow(i, j, -a / b);
+      if (inverse) tempInv.addToRow(i, j, -a / b);
+    }
+  }
+
+  // Now only diagonal elements are nonzero, but they are
+  // not necessarily 1. We normalise them.
+  for (unsigned i = 0; i < nRows; i++) {
+    a = m(i, i);
```

Comment:
this is not necessary. you can write m(i, i) directly below and remove the outer braces on these loops